modules/fedimint-ln-server/src/lib.rs

```diff
@@ -881,6 +882,13 @@ impl ServerModule for Lightning {
                         .await)
                 }
             },
+            api_endpoint! {
+                LIST_UNSPENT_INCOMING_CONTRACTS_ENDPOINT,
+                ApiVersion::new(0, 2),
+                async |module: &Lightning, context, _v: ()| -> Vec<ContractId> {
+                    Ok(module.list_unspent_incoming_contracts(&mut context.dbtx().into_nc()).await)
+                }
+            },
```

Comment:
Listing all unclaimed contracts isn't scalable at all. You can just use the account endpoint to query if a contract with a given contract id (which is the payment hash) exists. You just do that as long as you find contracts plus a gap limit.

There is a problem here that if more than gap limit many unpaid invoices were generated then you won't find later paid ones. This can be solved by introducing an API endpoint that lets you query incoming offers. Every time an invoice is generated an offer is created, so that doesn't depend on payment status.

Reply:
Hmm is a gap limit necessary? I thought the LNURL server would be able to keep track of the tweak indices that have been paid and be able to provide them to the client. Worst case scenario (if LNURL server went away) a wallet implementation could keep track of a gap limit and "last scanned index" themselves, but I thought it would be best to leave this up to higher level consumers of the client. That's why I added `indices: Vec<u64>`, to try to make the scanning a bit more efficient by only scanning the indices the LNURL server is tracking.

I agree that listing all unclaimed contracts isn't scalable, that was the main motivation with using the `LightningAuditItemKeyPrefix`, since I believe this should only query _funded_ unclaimed contracts. Unless there's a lot of concurrent lightning receives, I don't think this list should be too big. And these records are also cleaned up (unlike `ContractKey`), so the query should be performant. If you think its better to fetch each contract individually, we can change it to that.

Reply:
> Hmm is a gap limit necessary?

You are right, not necessary if you assume the LNURL server is reachable. A hybrid approach is probably best.